### ISSUE TITLE
false positive due to low-resolution Linux boot time

### ISSUE THREAD

**Issue by bugfood** [liske/needrestart#228]
Hi,

I'm running needrestart as a nagios check (`needrestart -p`) and I found a relatively rare instance of the check failing on a freshly-built server.

```
[Core] #7315 uses obsolete script file(s):
[Core] #7315  /usr/bin/<redacted>
```

I have high confidence that this is a false positive: the daemon could not have been running before the package which provided the daemon was installed; this was not a package upgrade.

The file in question has a `ctime` of:
```
$ stat -c "%.Z" /usr/bin/<redacted>
1638469345.021022179
```

Within `needrestart`, the `ctime` is handled at 1-second resolution, so that comes to `1638469345` (the low resolution here is not a problem for my use case--it could cause false negatives, but that's a separate issue).

The following test script isolates the code which `needrestart` uses to determine the process start time:
```
#!/usr/bin/perl

use strict;
use warnings;

use Proc::ProcessTable;

my %ptable = map {$_->pid => $_} @{ new Proc::ProcessTable(enable_ttys => 1)->table };
my $pid = 7315;
my $ps = $ptable{$pid};
print $ps->start, "\n";
```
This gives:
```
$ perl test.pl
1638469344
```

Compared to the file ctime, that value is lower by one, so `needrestart` returns a positive.

The reason this is happening is, as far as I can tell, due to a system limitation. On Linux, process start times are provided as time since the system boot, not as an absolute timestamp. In order to calculate an absolute process start time, `Proc::ProcessTable` must read the system boot time and then add the relative process start time. A further complication is that the process start time is given in clock ticks, so the number of ticks per second must be read. The full formula is thus:
```
absolute_process_start_time = boot_time + relative_process_start_time / ticks_per_second
```

The trouble comes from the resolution of the input values to that formula:
* boot time has a resolution of 1 second (the main problem)
* relative process start time has a resolution of 1 tick, hence 1/ticks_per_second seconds (a minor problem)

In my example case, boot time is `1638469201` (unix time) and relative process time is `14390` (ticks).
Thus the process start time is:
```
1638469201 + 14390 / 100 == 1638469344.9
```
Given the 1 second resolution of the boot time and the 1/100 second resolution of the relative process start time, that gives a range of:
```
1638469344.9 <= process_start_time < 1638469345.91
```

The file timestamp of `1638469345.02` falls well within that range.

I don't have a great solution to suggest for how to solve this. For my use case, I'd rather have `needrestart` tolerate values that are closer than 2 seconds (1.01 rounded up), though that would trade false positives for false negatives and I could imagine other use cases which are averse to that. My only suggestion is to allow a configurable tolerance, defaulting to 0 in order to avoid changing the current default behavior.

References:
* `proc(5)` (search for `starttime`) 
* `sysconf(3)` (search for `_SC_CLK_TCK`)
* https://metacpan.org/release/JWB/Proc-ProcessTable-0.634/source/os/Linux.c#L75
* https://metacpan.org/release/JWB/Proc-ProcessTable-0.634/source/os/Linux.c#L639

Thanks,
Corey

**Comment by liske** [comment]
Thanks for your deep look at this issue. Adding a configuration variable adding some tolerance sounds like the best option - the user can decide which trade-off works for him.

**Comment by bugfood** [comment]
Thanks, sounds good, in lieu of any better idea.

We've had some more occurrences of this, so I was able to put together a patch (linked above) and test it on an affected host.

**Comment by liske** [comment]
fixed by #233
